### PR TITLE
CSHARP-2509: Support Dictionary.ContainsValue in LINQ queries.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using MongoDB.Bson.Serialization.Attributes;
+using MongoDB.Bson.Serialization.Options;
+using MongoDB.Driver.Linq;
+using MongoDB.TestHelpers.XunitExtensions;
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Linq.Linq3Implementation.Jira
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:26]: The most interesting parts to look at in this PR are what MQL got generated. Is it the best possible MQL?

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
+                exception.Should().BeOfType<ArgumentException>();
+            }
+            else
+            {
+                var stages = Translate(collection, queryable);
+                AssertStages(stages, "{ $match : { $expr : { $reduce : { input : { $objectToArray : '$D1' }, initialValue : false, in : { $cond : { if : '$$value', then : true, else : { $eq : ['$$this.v', 1] } } } } } } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:48]: I couldn't figure out any way to support this without using `$expr`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
+                exception.Should().BeOfType<ArgumentException>();
+            }
+            else
+            {
+                var stages = Translate(collection, queryable);
+                AssertStages(stages, "{ $match : { $expr : { $reduce : { input : '$D2', initialValue : false, in : { $cond : { if : '$$value', then : true, else : { $eq : [{ $arrayElemAt : ['$$this', 1] }, 1] } } } } } } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:73]: I couldn't figure out any way to support this without using `$expr`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
+                exception.Should().BeOfType<ArgumentException>();
+            }
+            else
+            {
+                var stages = Translate(collection, queryable);
+                AssertStages(stages, "{ $match : { D3 : { $elemMatch : { v : 1 } } } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:98]: ArrayOfDocuments is the only representation that I could figure out a way to translate ContainsValue to MQL filter language without using `$expr`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
+                exception.Should().BeOfType<NotSupportedException>();
+            }
+            else
+            {
+                var stages = Translate(collection, queryable);
+                AssertStages(stages, "{ $project : { _v : { $reduce : { input : { $objectToArray : '$D1' }, initialValue : false, in : { $cond : { if : '$$value', then : true, else : { $eq : ['$$this.v', 1] } } } } }, _id : 0 } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:123]: Is there any better MQL we could translate to?

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
+                exception.Should().BeOfType<NotSupportedException>();
+            }
+            else
+            {
+                var stages = Translate(collection, queryable);
+                AssertStages(stages, "{ $project : { _v : { $reduce : { input : '$D2', initialValue : false, in : { $cond : { if : '$$value', then : true, else : { $eq : [{ $arrayElemAt : ['$$this', 1] }, 1] } } } } }, _id : 0 } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:148]: Is there any better MQL we could translate to?

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+            if (linqProvider == LinqProvider.V2)
+            {
+                var exception = Record.Exception(() => Translate(collection, queryable));
+                exception.Should().BeOfType<NotSupportedException>();
+            }
+            else
+            {
+                var stages = Translate(collection, queryable);
+                AssertStages(stages, "{ $project : { _v : { $reduce : { input : '$D3', initialValue : false, in : { $cond : { if : '$$value', then : true, else : { $eq : ['$$this.v', 1] } } } } }, _id : 0 } }");
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp2509Tests.cs:173]: Is there any better MQL we could translate to?

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request closed without merging**